gradle/publishing.gradle

```diff
@@ -0,0 +1,165 @@
+import com.android.build.api.artifact.ArtifactType
+import com.linkedin.gradle.FindApkTask
+
+apply plugin: 'maven-publish'
+apply plugin: 'signing'
+
+// Ideally AGP should provide sources and javadoc integration for their components:
+// https://issuetracker.google.com/issues/145670440
+task sourcesJar(type: Jar) {
+    classifier 'sources'
+    from android.sourceSets.main.java.srcDirs
+}
+
+task javadoc(type: Javadoc) {
+    source = android.sourceSets.main.java.srcDirs
+    if (project.plugins.hasPlugin("com.android.application")) {
+        android.applicationVariants.matching { variant -> variant.name == 'debug' }.all { variant ->
+            classpath += project.files(android.getBootClasspath().join(File.pathSeparator))
+            classpath += variant.getCompileClasspath(null)
+        }
+    }
+    if (project.plugins.hasPlugin("com.android.library")) {
+        android.libraryVariants.matching { variant -> variant.name == 'debug' }.all { variant ->
+            classpath += project.files(android.getBootClasspath().join(File.pathSeparator))
+            classpath += variant.getCompileClasspath(null)
+        }
+    }
+}
+
+task javadocJar(type: Jar, dependsOn: javadoc) {
+    classifier 'javadoc'
+    from javadoc.destinationDir
+}
+
+// AGP creates the components in afterEvaluate, so we need to use it too
```

Comment:
That's unfortunate but not entirely surprising; most configuration of publications must be done in an `afterEvaluate` block anyway when creation and configuration happen at different times.

Reply:
There is an interesting case that we found we need nested `afterEvaluate` 😭  https://github.com/detekt/detekt/pull/3453#discussion_r573472743